packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -177,15 +165,17 @@ function resetComponentStateWhenRemoved(vm: VM) {
         for (const key in oar) {
             oar[key].reset();
         }
-        runDisconnectedCallback(vm);
-        // Spec: https://dom.spec.whatwg.org/#concept-node-remove (step 14-15)
-        runShadowChildNodesDisconnectedCallback(vm);
-        runLightChildNodesDisconnectedCallback(vm);
+    }
+    if (isFalse(vm.isDirty)) {
+        // this guarantees that if the component is reused/reinserted,
+        // it will be re-rendered because we are disconnecting the reactivity
+        // linking, so mutations are not automatically reflected on the state
+        // of disconnected components.
```

Comment:
Going through the code, I think it means
```        
// this guarantees that if the component is reused/reinserted, it will be re-rendered.
// it should be re-rendered because we are disconnecting the reactivity
// linking, so mutations are not automatically reflected on the state
// of disconnected components.
```
@caridy can confirm.

Note: This piece of code was moved here from runDisconnectedCallback()